File: ocsgw/src/main/java/com/telenordigital/ocsgw/utils/DiameterUtilities.java
package com.telenordigital.ocsgw.utils;

import org.apache.log4j.Logger;
import org.jdiameter.api.Avp;
import org.jdiameter.api.AvpDataException;
import org.jdiameter.api.AvpSet;
import org.jdiameter.api.validation.Dictionary;
import org.jdiameter.common.impl.validation.DictionaryImpl;

public class DiameterUtilities {

    private static final Logger logger = Logger.getLogger(DiameterUtilities.class);

    private static final Dictionary AVP_DICTIONARY = DictionaryImpl.INSTANCE;

    public static void printAvps(AvpSet avps) {
        printAvps(avps, "");
    }

    private static void printAvps(AvpSet avps, String indentation) {
        for (Avp avp : avps) {
            String name = AVP_DICTIONARY.getAvp(avp.getCode(), avp.getVendorId()).getName();
            Object avpValue = getAvpValue(avp);
            StringBuilder avpLine = new StringBuilder(indentation + avp.getCode() + ": " + name);
            while (avpLine.length() < 50) {
                avpLine.append(avpLine.length() % 2 == 0 ? "." : " ");
            }
            avpLine.append(avpValue);
            logger.info(avpLine.toString());
            if (isGrouped(avp)) {
                try {
                    printAvps(avp.getGrouped(), indentation + "  ");
                } catch (AvpDataException e) {
                    // Failed to ungroup... ignore then...
                }
            }
        }
    }

    private static Object getAvpValue(Avp avp) {
        Object avpValue;
        try {
            String avpType = AVP_DICTIONARY.getAvp(avp.getCode(), avp.getVendorId()).getType();

            switch (avpType) {
                case "Integer32":
                case "AppId":
                    avpValue = avp.getInteger32();
                    break;
                case "Unsigned32":
                case "VendorId":
                    avpValue = avp.getUnsigned32();
                    break;
                case "Float64":
                    avpValue = avp.getFloat64();
                    break;
                case "Integer64":
                    avpValue = avp.getInteger64();
                    break;
                case "Time":
                    avpValue = avp.getTime();
                    break;
                case "Unsigned64":
                    avpValue = avp.getUnsigned64();
                    break;
                case "Grouped":
                    avpValue = "<Grouped>";
                    break;
                default:
                    avpValue = avp.getUTF8String().replaceAll("\r", "").replaceAll("\n", "");
                    break;
            }
        } catch (Exception ignore) {
            try {
                avpValue = avp.getUTF8String().replaceAll("\r", "").replaceAll("\n", "");
            } catch (AvpDataException e) {
                avpValue = avp.toString();
            }
        }
        return avpValue;
    }

    private static boolean isGrouped(Avp avp) {
        boolean grouped = false;

        try {
            String avpType = AVP_DICTIONARY.getAvp(avp.getCode(), avp.getVendorId()).getType();

            if ("Grouped".equals(avpType)) {
               grouped = true;
            }
        } catch (Exception ignore) {
            // Not grouped
        }
        return  grouped;
    }
}

File: diameter-stack/src/main/kotlin/com/telenordigital/ostelco/diameter/util/DiameterUtilities.kt
package com.telenordigital.ostelco.diameter.util

import com.telenordigital.ostelco.diameter.logger
import org.jdiameter.api.Avp
import org.jdiameter.api.AvpDataException
import org.jdiameter.api.AvpSet
import org.jdiameter.common.impl.validation.DictionaryImpl

class DiameterUtilities {

    private val LOG by logger()

    private val AVP_DICTIONARY = DictionaryImpl.INSTANCE

    fun printAvps(avps: AvpSet?) {
        if (avps != null) {
            printAvps(avps, "")
        }
    }

    private fun printAvps(avps: AvpSet, indentation: String) {
        for (avp in avps) {
            val name = AVP_DICTIONARY.getAvp(avp.code, avp.vendorId).name
            val avpValue = getAvpValue(avp)
            val avpLine = StringBuilder(indentation + avp.code + ": " + name)
            while (avpLine.length < 50) {
                avpLine.append(if (avpLine.length % 2 == 0) "." else " ")
            }
            avpLine.append(avpValue)
            LOG.info(avpLine.toString())
            if (isGrouped(avp)) {
                try {
                    printAvps(avp.grouped, "$indentation  ")
                } catch (e: AvpDataException) {
                    // Failed to ungroup... ignore then...
                }

            }
        }
    }

    private fun getAvpValue(avp: Avp): Any {
        var avpValue: Any
        try {
            val avpType = AVP_DICTIONARY.getAvp(avp.code, avp.vendorId).type

            when (avpType) {
                "Integer32", "AppId" -> avpValue = avp.integer32
                "Unsigned32", "VendorId" -> avpValue = avp.unsigned32
                "Float64" -> avpValue = avp.float64
                "Integer64" -> avpValue = avp.integer64
                "Time" -> avpValue = avp.time
                "Unsigned64" -> avpValue = avp.unsigned64
                "Grouped" -> avpValue = "<Grouped>"
                else -> avpValue = (avp.utF8String as String)
                        .replace("\r", "")
                        .replace("\n", "")
            }
        } catch (ignore: Exception) {
            try {
                avpValue = avp.utF8String
                        .replace("\r", "")
                        .replace("\n", "")
            } catch (e: AvpDataException) {
                avpValue = avp.toString()
            }

        }
        return avpValue
    }

    // TODO for missing Avp, is code and vendorId as 0 okay?
    private fun isGrouped(avp: Avp?): Boolean =
            ("Grouped" == AVP_DICTIONARY.getAvp(avp?.code ?: 0, avp?.vendorId ?: 0).type)
}
